### DISCUSSION TITLE
Inline visualtions of vedo plots in ipynb

### DISCUSSION THREAD

**Discussion by clemens-fricke** [tataratat/gustaf#110] [Ideas]
Das ist nur ein nice to have und was Jae und ich schon durch trial und error herrausgefunden haben, funktioniert es auch nicht perfekt out of the box.

Wenn man das richtige vedo backend eingestellt hat (ipyvtk) *diese funktionalität müsste auch noch hinzugefügt werden* muss bei show das keyword '''close=False''' gesetzt werden und dann auf dem plotter der dann zurück gegeben wird noch mal show() ausführen. 

What needs to be done
==================

- Add settings option for vedo backend
     * gustaf.vedo.settings.VEDO_BACKEND = None
     * Optional VEDO_BACKEND = "ipyvtk"
- Either check if VEDO_BACKEND == "ipyvtk" than always return plot without closing it or give warning that close=false should be set so that .show() can be called on the return value inside the ipynb.

**Comment by clemens-fricke** [comment]
After some more investigations, I could not get it to work in the same manner jae and I had while we tried it as a spike.

A consistent way to display interactive vedo graphics inline in ipynb is to create the plotter in the notebook and to collect the vedo objects by calling gus_obj.showable() and calling the plotter directly with the returned objects. See example below.

```
import gustaf as gus
gus.utils.log.configure(debug=True)

# create gus example object
v_res = [20, 30, 40]
vertices = gus.create.vertices.raster(bounds=[[-1, 0, 5], [4, 1, 10]], resolutions=v_res)
connec = gus.utils.connec.make_hexa_volumes(v_res)
v = gus.Volumes(vertices.vertices, connec)

# collect showables (vedo objects)
a = v.showable()

# create the plotter with the correct backend
plt = Plotter(backend="ipyvtk")

# call the show function directly on the plotter with the needed objects
plt.show(a, at=0, interactive=True, camera=None)

# close the plotter to not create a memory leak
# this can be in another cell to keep the figure interactive until you are done with it
plt.close()
```

**Comment by j042** [comment]
> * Either check if VEDO_BACKEND == "ipyvtk" than always return plot without closing it or give warning that close=false should be set so that .show() can be called on the return value inside the ipynb.

Sounds good. I'd say we can add a function `notebook_plotter` (or some other name) that wraps `show_vedo`.
```python
...
plt = gus.show.notebook_plotter([toshow1, toshow2])
plt.show()
```
in case of multiple plots at once:
```python
>>> plt1, plt2, plt3 = gus.show.notebook_plotter(toshow1, toshow2, toshow3)
>>> plotters = gus.show.notebook_plotter(toshow1, toshow2, toshow3)
>>> isinstance(plotters, list)
True
```

**Comment by j042** [comment]
closing for now

**Comment by j042** [comment]
@clemens-fricke I think we are ready to implement this

**Comment by clemens-fricke** [comment]
Quick update.

Vedo changed how the backend is handled and you can not change it anymore in the plotter call directly. But you can set it globally with `vedo.settings.backend = "..."`

Konstantin and I just played around with it and we got it to work with:

```
import vedo
import gustaf as gus
vedo.settings.default_backend="k3d"
plt = vedo.Plotter()
```
The following actually does not work, because k3d can not plot UGrid which is apparently how we export the volume element.
```
# create gus example object
v_res = [20, 30, 40]
vertices = gus.create.vertices.raster(bounds=[[-1, 0, 5], [4, 1, 10]], resolutions=v_res)
connec = gus.utils.connec.make_hexa_volumes(v_res)
v = gus.Volumes(vertices.vertices, connect)
```
What works is the plotting of 3D Faces...
```
V = np.array(
        [
            [0.0, 0.0, 0.0],
            [1.0, 0.0, 0.0],
            [0.0, 1.0, 0.0],
            [1.0, 1.0, 0.0],
            [0.0, 0.0, 1.0],
            [1.0, 0.0, 1.0],
            [0.0, 1.0, 1.0],
            [1.0, 1.0, 1.0],
        ],
        dtype=np.float64,
    )
QF = np.array(
        [
            [1, 0, 2, 3],
            [0, 1, 5, 4],
            [1, 3, 7, 5],
            [3, 2, 6, 7],
            [2, 0, 4, 6],
            [4, 5, 7, 6],
        ],
        dtype=np.int32,
    )
face = gus.Faces(V, QF)
```
```
plt.show(face.showable())
```


in regard to splinepy splines

showable returns a dicts to plot it you need to do the following:
```
plt.show(*list(spline.showable().values()))
```

**Comment by clemens-fricke** [comment]
On its way via k3d plotting. PR is open #181 but also needs #171 to be merged. In addition, if you want to use it in conjunction with splinepy the PR tataratat/splinepy#311 needs to be merged.

There might be some parts missing, please open an Issue if you find something you need but does not work with this.
